src/ray/rpc/retryable_grpc_client.h

```diff
@@ -0,0 +1,306 @@
+      request,                                                                \
+      callback,                                                               \
+      method_timeout_ms))
+
+/// \class Executor
+/// Executor saves operation and support retries.
+class Executor {
+ public:
+  Executor(std::function<void(const ray::Status &)> abort_callback)
+      : abort_callback_(std::move(abort_callback)) {}
+
+  /// This function is used to execute the given operation.
+  ///
+  /// \param operation The operation to be executed.
+  void Execute(std::function<void()> operation) {
+    operation_ = std::move(operation);
+    operation_();
+  }
+
+  /// This function is used to retry the given operation.
+  void Retry() { operation_(); }
+
+  void Abort(const ray::Status &status) { abort_callback_(status); }
+
+ private:
+  std::function<void(ray::Status)> abort_callback_;
+  std::function<void()> operation_;
+};
+
+class RetryableGrpcClient : public std::enable_shared_from_this<RetryableGrpcClient> {
+ public:
+  static std::shared_ptr<RetryableGrpcClient> Create(
+      std::shared_ptr<grpc::Channel> channel,
+      instrumented_io_context &io_context,
+      uint64_t max_pending_requests_bytes,
+      uint64_t check_channel_status_interval_milliseconds,
+      uint64_t server_unavailable_timeout_seconds,
+      std::function<void()> server_unavailable_timeout_callback,
+      const std::string &server_name) {
+    auto retryable_grpc_client = std::shared_ptr<RetryableGrpcClient>(
+        new RetryableGrpcClient(channel,
+                                io_context,
+                                max_pending_requests_bytes,
+                                check_channel_status_interval_milliseconds,
+                                server_unavailable_timeout_seconds,
+                                server_unavailable_timeout_callback,
+                                server_name));
+    // SetupCheckTimer() MUST be called after we have the shard_ptr of
+    // RetryableGrpcClient since it calls shared_from_this()
+    // which requires a shared_ptr of this to exist.
+    retryable_grpc_client->SetupCheckTimer();
+    return retryable_grpc_client;
+  }
+
+  RetryableGrpcClient(const RetryableGrpcClient &) = delete;
+  RetryableGrpcClient &operator=(const RetryableGrpcClient &) = delete;
+
+  template <typename Service, typename Request, typename Reply>
+  void CallMethod(PrepareAsyncFunction<Service, Request, Reply> prepare_async_function,
+                  GrpcClient<Service> &grpc_client,
+                  const std::string &call_name,
+                  const Request &request,
+                  const ClientCallback<Reply> &callback,
+                  const int64_t timeout_ms) {
+    auto executor = new Executor(
+        [callback](const ray::Status &status) { callback(status, Reply()); });
+    std::weak_ptr<RetryableGrpcClient> weak_self = shared_from_this();
```

Comment:
This weak_ptr handles the case when RetryableGrpcClient is already destructed, after which the 
 operation_callback is called. However we don't have same treatments to `&grpc_client` which faces similar issues. We need to think about the lifetimes of a single call, a retryable call, &grpc_client and RetryableGrpcClient.

Reply:
Remove `weak_ptr` and rely on `shutdown_`.